src/rust/engine/src/nodes.rs

```diff
@@ -205,7 +206,25 @@ impl From<Select> for NodeKey {
   }
 }
 
-pub fn lift_digest(digest: &Value) -> Result<hashing::Digest, String> {
+pub fn lift_directory_digest(types: &Types, digest: &Value) -> Result<hashing::Digest, String> {
+  if types.directory_digest != externs::get_type_for(digest) {
+    return Err(format!(
+      "{} is not of type {}.",
+      digest, types.directory_digest
+    ));
+  }
+  let fingerprint = externs::project_str(&digest, "fingerprint");
+  let digest_length = externs::project_u64(&digest, "serialized_bytes_length") as usize;
+  Ok(hashing::Digest(
+    hashing::Fingerprint::from_hex_string(&fingerprint)?,
+    digest_length,
+  ))
+}
+
+pub fn lift_file_digest(types: &Types, digest: &Value) -> Result<hashing::Digest, String> {
+  if types.file_digest != externs::get_type_for(digest) {
+    return Err(format!("{} is not of type {}.", digest, types.file_digest));
+  }
```

Comment:
A lot of the churn in this PR is due to the addition of these type checks.

I expect that in the medium term we should consider moving all of the `fs.py` types to Rust using `cpython`'s class generation mechanism, as it's likely both safer and more efficient than the `externs::project_*` boilerplate: https://github.com/pantsbuild/pants/blob/4129e0622d456be7b17493be6dda900894adb39c/src/rust/engine/src/externs/interface.rs#L463-L488